src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -51,51 +65,84 @@ public class AsyncSemaphore : IDisposable
-        /// <returns>A task whose result is a releaser that should be disposed to release the lock.</returns>
-        public Task<Releaser> EnterAsync(CancellationToken cancellationToken = default(CancellationToken))
-        {
-            if (cancellationToken.IsCancellationRequested)
-            {
-                return Task.FromCanceled<Releaser>(cancellationToken);
-            }
-
-            if (this.disposed)
-            {
-                return DisposedReleaserTask;
-            }
-
-            return this.LockWaitingHelper(this.semaphore.WaitAsync(cancellationToken));
-        }
+        /// <returns>
+        /// A task whose result is a releaser that should be disposed to release the lock.
+        /// This task may be canceled if <paramref name="cancellationToken"/> is signaled.
+        /// </returns>
+        /// <exception cref="OperationCanceledException">Thrown when <paramref name="cancellationToken"/> is canceled before semaphore access is granted.</exception>
+        /// <exception cref="ObjectDisposedException">Thrown when this semaphore is disposed before semaphore access is granted.</exception>
+        public Task<Releaser> EnterAsync(CancellationToken cancellationToken = default) => this.EnterAsync(Timeout.InfiniteTimeSpan, cancellationToken);
 
         /// <summary>
         /// Requests access to the lock.
         /// </summary>
         /// <param name="timeout">A timeout for waiting for the lock.</param>
         /// <param name="cancellationToken">A token whose cancellation signals lost interest in the lock.</param>
-        /// <returns>A task whose result is a releaser that should be disposed to release the lock.</returns>
-        public Task<Releaser> EnterAsync(TimeSpan timeout, CancellationToken cancellationToken = default(CancellationToken))
+        /// <returns>
+        /// A task whose result is a releaser that should be disposed to release the lock.
+        /// This task may be canceled if <paramref name="cancellationToken"/> is signaled or <paramref name="timeout"/> expires.
+        /// </returns>
+        /// <exception cref="OperationCanceledException">Thrown when <paramref name="cancellationToken"/> is canceled or the <paramref name="timeout"/> expires before semaphore access is granted.</exception>
+        /// <exception cref="ObjectDisposedException">Thrown when this semaphore is disposed before semaphore access is granted.</exception>
+        public Task<Releaser> EnterAsync(TimeSpan timeout, CancellationToken cancellationToken = default)
         {
-            if (this.disposed)
+            if (cancellationToken.IsCancellationRequested)
             {
-                return DisposedReleaserTask;
+                return Task.FromCanceled<Releaser>(cancellationToken);
             }
 
-            return this.LockWaitingHelper(this.semaphore.WaitAsync(timeout, cancellationToken));
+            lock (this.syncObject)
```

Comment:
Technically this prohibits a true FIFO guarantee, e.g. thread 1 is holding the lock, thread 2 comes along and waits on the lock, then thread 3 comes along and waits on the lock... when thread 1 releases the lock, thread 3 could be allowed in rather than thread 2.  If all you care about is the order of the linked list, though, then it's fine.

Reply:
I had no idea that Monitor didn't guarantee FIFO ordering. Is there a better alternative?
or, if this is good enough, is it no better than keeping with SemaphoreSlim?

Reply:
Well, this PR improves throughput:

fix502:
Completed 2614057 cycles
Completed 2580033 cycles

master:
Completed 1773521 cycles
Completed 1965890 cycles

Reply:
> I had no idea that Monitor didn't guarantee FIFO ordering

FWIW, pretty much no synchronization in .NET is fair.

> is it no better than keeping with SemaphoreSlim?

From the perspective of the specific guarantees you're hoping for, no, other than you controlling it and being able to reject PRs that might break it. :)  I can't 100% promise such an explicit ordering change would ever find its way into SemaphoreSlim, though I think it's unlikely.